### PR TITLE
Decouple from Scarb

### PR DESCRIPTION
**Pull request by piotmag769**
Related scarb PR: https://github.com/software-mansion/scarb/pull/886

## Introduced changes

- moved test collection and compilation to Scarb

## Checklist

- [x] Linked relevant issue
- [ ] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Comment by piotmag769** [comment]
Tests won't pass bcs they require Scarb with `snforge-test-collector` xD

File: crates/cheatnet/tests/cheatcodes/declare.rs [hunk 1]
```diff
@@ -7,9 +7,9 @@ use starknet_api::hash::StarkFelt;
 use starknet_api::stark_felt;
 
 static HELLO_STARKNET_CLASS_HASH: &str =
-    "0x029c29c28d7eed24e3e57e6189cda4fc5a7a5f1d097f02c6692cff91816a5fbc";
+    "0x04ac1b5bcf427bef9eb31691de4104268179e26df6e81eb93492870c3e7ca9cb";
 static CONSTRUCTOR_SIMPLE_CLASS_HASH: &str =
-    "0x045b3e15b898999ca953f197065fb4b4f98af6e92fb012bb26917a8020c3314d";
+    "0x06aa43ff1d4a13360a7ec9ab34924a735adee7fd42042b8d4f910d8386dce556";
```
Comment (cptartur) [inline crates/cheatnet/tests/cheatcodes/declare.rs:12]: Why did these change?

File: crates/forge/src/collecting.rs
Comment (cptartur) [inline crates/forge/src/collecting.rs:1]: This file should probably be renamed

File: crates/forge/src/lib.rs [hunk 2]
```diff
@@ -216,21 +209,22 @@ fn to_runnable(
...
 /// * `contracts` - Map with names of contract used in tests and corresponding sierra and casm artifacts
-/// * `predeployed_contracts` - Absolute path to predeployed contracts used by starknet state e.g. account contracts
 ///
 
 #[allow(clippy::implicit_hasher)]
 pub async fn run(
-    package_path: &Utf8Path,
     package_name: &str,
-    package_source_dir_path: &Utf8Path,
+    snforge_target_dir_path: &Utf8Path,
```
Comment (cptartur) [inline crates/forge/src/lib.rs:217]: Why not just `target_dir_path`

Comment (cptartur) [reply crates/forge/src/lib.rs:217]: It's in the context of this crate anyways.

File: crates/forge/src/main.rs [hunk 3]
```diff
@@ -166,6 +167,17 @@ fn test_workspace(args: TestArgs) -> Result<bool> {
...
+    let test_build_output = Command::new("scarb")
+        .arg("snforge-test-collector")
+        .env("SCARB_PACKAGES_FILTER", filter_env)
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .output()
+        .context("Failed to build test artifacts with Scarb")?;
+    if !test_build_output.status.success() {
+        bail!("scarb snforge-test-collector did not succeed")
+    }
```
Comment (Arcticae) [inline crates/forge/src/main.rs:179]: this deserves own function

File: crates/forge/src/test_filter.rs [hunk 4]
```diff
@@ -99,8 +98,9 @@ mod tests {
     use crate::collecting::CompiledTestCrate;
     use crate::test_filter::TestsFilter;
     use crate::CrateLocation;
+    use crate::TestCase;
     use cairo_lang_sierra::program::Program;
-    use test_collector::{ExpectedTestResult, TestCase};
+    use cairo_lang_test_plugin::test_config::TestExpectation;
```
Comment (Arcticae) [inline crates/forge/src/test_filter.rs:103]: How is this change related exactly?

File: crates/forge/src/lib.rs [hunk 5]
```diff
@@ -216,21 +209,22 @@ fn to_runnable(
...
     runner_params: Arc<RunnerParams>,
 ) -> Result<Vec<TestCrateSummary>> {
-    let compilation_targets =
-        collect_test_compilation_targets(package_path, package_name, package_source_dir_path)?;
-    let test_crates = compile_tests(&compilation_targets, &runner_params)?;
+    let snforge_test_artifact_path =
+        snforge_target_dir_path.join(format!("{package_name}.sierra.json"));
+    let test_crates = serde_json::from_str::<Vec<CompiledTestCrateRaw>>(&std::fs::read_to_string(
+        snforge_test_artifact_path,
+    )?)?;
```
Comment (cptartur) [inline crates/forge/src/lib.rs:226]: If the artefacts compilation is going on in scarb now, perhaps this function should already accept them in the deserialized (as structs/vec/whatever rust object) instead of taking the path. 

This way we can still use it in integration tests in an easier way.

File: crates/forge/test_utils/src/runner.rs [hunk 6]
```diff
@@ -99,6 +105,33 @@ impl<'a> TestCase {
 
         dir.child("src/lib.cairo").touch().unwrap();
 
+        let snforge_std_path = Utf8PathBuf::from_str("../../snforge_std")
+            .unwrap()
+            .canonicalize_utf8()
+            .unwrap()
+            .to_string()
+            .replace('\\', "/");
```
Comment (cptartur) [inline crates/forge/test_utils/src/runner.rs:113]: Why are we replacing this?

File: crates/forge/tests/integration/dispatchers.rs [hunk 7]
```diff
@@ -576,7 +576,7 @@ fn nonexistent_method_call() {
     assert_case_output_contains!(
         result,
         "test_nonexistent_method_call",
-        "Entry point selector 0x01fdb214e1495025fa4baf660d34f03c0d8b5037cf10311d2a3202a806aa9485 not found in contract 0x00c0727f111a1f9fd7543cf36ae25feaa8184a9188daa4c9c148605445dcba9f"
+        "Entry point selector 0x01fdb214e1495025fa4baf660d34f03c0d8b5037cf10311d2a3202a806aa9485 not found in contract 0x0612c9ea6cdc79914ba9eae055225ee9df90368b02c9e463247d6c2d313000bb"
```
Comment (Arcticae) [inline crates/forge/tests/integration/dispatchers.rs:579]: What did cause that bump?

File: crates/forge/tests/e2e/running.rs [hunk 8]
```diff
@@ -614,6 +614,7 @@ fn init_new_project_test() {
             [package]
             name = "test_name"
             version = "0.1.0"
+            edition = "2023_10"
```
Comment (Arcticae) [inline crates/forge/tests/e2e/running.rs:617]: Is this now a required parameter?

File: crates/forge/test_utils/src/runner.rs [hunk 9]
```diff
@@ -99,6 +105,33 @@ impl<'a> TestCase {
...
+                [package]
+                name = "test_package"
+                version = "0.1.0"
+    
+                [[target.starknet-contract]]
+                sierra = true
+                casm = true
+    
+                [dependencies]
+                starknet = "2.3.1"
```
Comment (cptartur) [inline crates/forge/test_utils/src/runner.rs:128]: This will have to be manually updated with every cairo bump which we'll probably forget about. Perhaps this value should be taken from top-level `.tool-versions` file, assuming scarb will have the same versions as starknet.

File: crates/forge/tests/integration/spy_events.rs [hunk 10]
```diff
@@ -428,7 +428,7 @@ fn event_emitted_wrong_data_asserted() {
     assert_case_output_contains!(
         result,
         "test_assert_wrong_data",
-        "2814149079029271315825228546613746536844456515452335425622822705028162471071"
+        "385731233638135702203933925049152883559461865803221708830049339625617856128"
```
Comment (cptartur) [inline crates/forge/tests/integration/spy_events.rs:431]: Why did it change?

File: crates/scarb-artifacts/src/lib.rs [hunk 11]
```diff
@@ -164,24 +171,20 @@ pub fn corelib_for_package(metadata: &Metadata, package: &PackageId) -> Result<U
...
-
     let package = metadata
         .get_package(package)
         .ok_or_else(|| anyhow!("Failed to find metadata for package = {package}"))?;
 
-    let package_path = package.root.clone();
-    let package_source_dir_path = compilation_unit.target.source_root();
-
-    Ok((package_path, Utf8PathBuf::from(package_source_dir_path)))
+    Ok((package.root.clone(), package.manifest_path.clone()))
```
Comment (cptartur) [inline crates/scarb-artifacts/src/lib.rs:183]: Was this only used in forge? So we don't accidentally break some behavior in cast.

File: crates/forge/tests/e2e/running.rs [hunk 12]
```diff
@@ -759,12 +760,10 @@ fn available_gas_error() {
     let snapbox = test_runner();
 
     let output = snapbox.current_dir(&temp).assert().failure();
-    assert_stdout_contains!(
+    assert_stderr_contains!(
```
Comment (cptartur) [inline crates/forge/tests/e2e/running.rs:763]: Why did it only change here while other diagnostics are still printed to stdout?

File: crates/cheatnet/tests/cheatcodes/declare.rs (hunk 1 above)
Comment (drknzz) [reply crates/cheatnet/tests/cheatcodes/declare.rs:12]: +1, although these are computed automatically on master ([#1104](https://github.com/foundry-rs/starknet-foundry/pull/1104))

File: crates/forge/src/collecting.rs [hunk 13]
```diff
@@ -36,169 +35,34 @@ pub(crate) struct TestCompilationTarget {
...
-    let tests_dir_path = package_path.join("tests");
-    if tests_dir_path.exists() {
-        compilation_targets.push(TestCompilationTarget {
-            crate_name: "tests".to_string(),
-            crate_location: CrateLocation::Tests,
-            lib_content: get_or_create_test_lib_content(tests_dir_path.as_path())?,
-            crate_root: tests_dir_path,
-        });
-    }
+pub trait ForkConfig {}
```
[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/src/collecting.rs:38]: Why? It was a conscious decision, it is copied from somewhere else. It is to parametrisation of `TestCase<T>` with random type - the type has to implement this empty trait

File: crates/forge/src/collecting.rs [hunk 14]
```diff
@@ -36,169 +35,34 @@ pub(crate) struct TestCompilationTarget {
...
 
-    Ok(compilation_targets)
-}
+impl ForkConfig for RawForkConfig {}
```
[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/src/collecting.rs:40]: Above^

File: crates/forge/src/lib.rs [hunk 15]
```diff
@@ -216,21 +209,22 @@ fn to_runnable(
 /// * `runner_config` - A configuration of the test runner
 /// * `corelib_path` - Absolute path to the Cairo corelib
 /// * `contracts` - Map with names of contract used in tests and corresponding sierra and casm artifacts
-/// * `predeployed_contracts` - Absolute path to predeployed contracts used by starknet state e.g. account contracts
 ///
 
 #[allow(clippy::implicit_hasher)]
 pub async fn run(
```
[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/src/lib.rs:215]: You mean the comment? Agree, will do

File: crates/forge/src/main.rs (hunk 3 above)
Comment (piotmag769) [reply crates/forge/src/main.rs:179]: Then we want `scarb build` in its own function too?

File: crates/forge/src/lib.rs (hunk 2 above)
Comment (piotmag769) [reply crates/forge/src/lib.rs:217]: Disagree, with `target_dir_path` you don't really know if it is a path to `target` or `target/dev/snforge`

File: crates/forge/test_utils/src/runner.rs (hunk 6 above)
Comment (maciektr) [reply crates/forge/test_utils/src/runner.rs:113]: `\\` are used on windows, this unifies paths across systems

File: crates/forge/tests/e2e/running.rs (hunk 8 above)
Comment (maciektr) [reply crates/forge/tests/e2e/running.rs:617]: It's optional, if not filled in it defaults to an edition set as default on compiler side - currently `2023_01`

File: crates/forge/test_utils/src/runner.rs (hunk 9 above)
Comment (piotmag769) [reply crates/forge/test_utils/src/runner.rs:128]: Out of scope + we should stop updating it manually in all the other places anyways, it should work until a new Cairo (and thus Scarb) major is released.

Imagine if we did not bump it in `snforge_std` for 2.3.1. You will have `starknet=2.3.0` specified, which is the same as `starknet=^2.3.0`, so it would still work with both Scarb 2.3.1 and 2.3.0. Now we basically limited our users to use Scarb 2.3.1 only.

Summary: until a new major of Scarb/Cairo is released, updating starknet dependency in Scarb.tomls for tests is not harmful, but useless. Updating it in `snforge_std` is limiting the range of compatible Scarb versions (we should stop doing that).

File: crates/forge/src/lib.rs (hunk 5 above)
Comment (piotmag769) [reply crates/forge/src/lib.rs:226]: This would put even more logic into `main.rs`, I don't really like that. Also it doesn't simplify integration tests imo, we don't want to generate them manually for tests anyways

File: crates/forge/tests/e2e/running.rs (hunk 8 above)
Comment (piotmag769) [reply crates/forge/tests/e2e/running.rs:617]: If not specified it uses the default, but this is the expected toml content. We use `scarb new` which generates it in the Scarb.toml of the new package

File: crates/forge/tests/integration/dispatchers.rs (hunk 7 above)
Comment (piotmag769) [reply crates/forge/tests/integration/dispatchers.rs:579]: I use the Scarb built locally for testing, it has a different compiler version

File: crates/cheatnet/tests/cheatcodes/declare.rs (hunk 1 above)
Comment (piotmag769) [reply crates/cheatnet/tests/cheatcodes/declare.rs:12]: https://github.com/foundry-rs/starknet-foundry/pull/1121#discussion_r1394072646

Comment (piotmag769) [reply crates/cheatnet/tests/cheatcodes/declare.rs:12]: Will rebase on master

File: crates/forge/tests/integration/spy_events.rs (hunk 10 above)
Comment (piotmag769) [reply crates/forge/tests/integration/spy_events.rs:431]: https://github.com/foundry-rs/starknet-foundry/pull/1121#discussion_r1394072646

File: crates/scarb-artifacts/src/lib.rs (hunk 11 above)
Comment (piotmag769) [reply crates/scarb-artifacts/src/lib.rs:183]: Yup, it was only used in forge

File: crates/forge/tests/e2e/running.rs (hunk 12 above)
Comment (piotmag769) [reply crates/forge/tests/e2e/running.rs:763]: It is not a diagnostic, it is a custom error of ours. It comes from stderr of `scarb-snforge-test-collector` which we inherit.

In case of erros caught directly in forge we print them `fn main` in `forge/src/main.rs` to the stdout (don't really know why, I wanted to change it someday but remember we decided not to do it).

File: crates/forge/test_utils/src/runner.rs (hunk 9 above)
Comment (maciektr) [reply crates/forge/test_utils/src/runner.rs:128]: Yes, https://docs.swmansion.com/scarb/docs/reference/specifying-dependencies.html#version-requirements

File: crates/forge/test_utils/src/runner.rs (hunk 6 above)
Comment (maciektr) [reply crates/forge/test_utils/src/runner.rs:113]: A comment would be nice though

File: crates/forge/src/main.rs [hunk 16]
```diff
@@ -166,6 +167,17 @@ fn test_workspace(args: TestArgs) -> Result<bool> {
         bail!("Scarb build did not succeed")
     }
 
+    let test_build_output = Command::new("scarb")
```
Comment (maciektr) [inline crates/forge/src/main.rs:170]: If you want to rely on scarb profile chose by user (line 149), you should also pass it to scarb on build cmd via `SCARB_PROFILE` env var. https://docs.swmansion.com/scarb/docs/writing-extensions/subcommands.html#environment-variables

File: crates/forge/src/test_filter.rs (hunk 4 above)
Comment (piotmag769) [reply crates/forge/src/test_filter.rs:103]: Oh wait, those shouldn't be here, left from the first version xD Thx

File: crates/forge/src/lib.rs (hunk 5 above)
Comment (cptartur) [reply crates/forge/src/lib.rs:226]: This doesn't necessarily have to be in main, it can be a separate function in lib that would be called before `run`. Similarly to how collecting and compilation was split into multiple functions called in a sequence

File: crates/forge/test_utils/src/runner.rs (hunk 9 above)
Comment (cptartur) [reply crates/forge/test_utils/src/runner.rs:128]: Ideally we'd test it across some range of values, because this approach only ensures forge works with a specific version. Even if, in theory, it should work with multiple.

File: crates/forge/src/lib.rs (hunk 2 above)
Comment (cptartur) [reply crates/forge/src/lib.rs:217]: It's not immediately obvious that it is meant to be a path to `target/dev/snforge` with this name either. Target dir could also suggest a directory for which you want to run tests i.e. scarb package.

File: crates/forge/tests/e2e/running.rs (hunk 12 above)
Comment (cptartur) [reply crates/forge/tests/e2e/running.rs:763]: It could be a weird from the perspective of use to show these errors like this, without an `[ERROR]` tag before the message or whatever. 

It may be fine if it's just `available_gas` but we should consider some approach of displaying the collector errors in manner more consistent with the rest of the forge.

File: crates/forge/src/lib.rs (hunk 5 above)
Comment (piotmag769) [reply crates/forge/src/lib.rs:226]: I think we just want to replace the collection + compilation logic here, no? So it would make sense to read it here I think, I can extract it to a separate function that with a docstring explaining what it does

Comment (cptartur) [reply crates/forge/src/lib.rs:226]: This seems like a good middle ground for now, we can do that.

File: crates/forge/src/main.rs (hunk 16 above)
Comment (piotmag769) [reply crates/forge/src/main.rs:170]: I guess every if every scarb process uses this env var it should be fine

Comment (maciektr) [reply crates/forge/src/main.rs:170]: Yes, we do

File: crates/forge/tests/e2e/running.rs (hunk 12 above)
Comment (piotmag769) [reply crates/forge/tests/e2e/running.rs:763]: Minor thing imo, you can add an issue if you want

Comment (cptartur) [reply crates/forge/tests/e2e/running.rs:763]: Quite a change in UX in my opinion, and annotations and errors emitted by them are already limited.

File: crates/forge/src/lib.rs (hunk 2 above)
Comment (piotmag769) [reply crates/forge/src/lib.rs:217]: I think it's obvious what target directory is considering we are in Rust-like environment

File: crates/forge/tests/e2e/diagnostics_and_plugins.rs [hunk 17]
```diff
@@ -80,11 +80,15 @@ fn print_error_if_attributes_incorrect() {
         #[fuzzer()]
                 ^^
         
-        error: Plugin diagnostic: Expected panic must be of the form `expected: <tuple of felt252s>`.
+        error: Plugin diagnostic: Expected panic must be of the form `expected: <tuple of felt252s and strings>` or `expected: "some string"` or `expected: <some felt252>`.
+         --> {mock_tests_dir_path_str}/tests/test_should_panic.cairo:2:15
+        #[should_panic(url: "https://test.com")]
+                      ^***********************^
+        
+        error: Plugin diagnostic: Expected panic must be of the form `expected: <tuple of felt252s and strings>` or `expected: "some string"` or `expected: <some felt252>`.
```
Comment (piotmag769) [inline crates/forge/tests/e2e/diagnostics_and_plugins.rs:88]: #1128

File: CHANGELOG.md [hunk 18]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.11.0-rc.0] - 2023-11-17
+
```
Comment (piotmag769) [inline CHANGELOG.md:11]: Will be reverted, made release for testing purposes

File: Cargo.toml [hunk 19]
```diff
@@ -2,15 +2,14 @@
...
-    "crates/test-collector",
     "crates/cast",
     "crates/cheatnet",
     "crates/conversions",
     "crates/scarb-artifacts",
 ]
 
 [workspace.package]
-version = "0.10.2"
+version = "0.11.0-rc.0"
```
Comment (piotmag769) [inline Cargo.toml:12]: Same here ^

File: crates/forge/src/lib.rs [hunk 20]
```diff
@@ -206,31 +199,38 @@ fn to_runnable(
...
+
 /// Run the tests in the package at the given path
 ///
 /// # Arguments
 ///
-/// * `package_path` - Absolute path to the top-level of the Cairo package
-/// * `lib_path` - Absolute path to the main file in the package (usually `src/lib.cairo`)
-/// * `linked_libraries` - Dependencies needed to run the package at `package_path`
+/// * `package_name` - Name of the package specified in Scarb.toml
+/// * `snforge_target_dir_path` - Absolute path to the directory with snforge test artifacts (`{package_path}/target/{profile_name}/snforge`)
```
Comment (cptartur) [inline crates/forge/src/lib.rs:220]:
```suggestion
/// * `snforge_target_dir_path` - Absolute path to the directory with snforge test artifacts (Usually `{package_path}/target/{profile_name}/snforge`)
```

File: crates/forge/test_utils/src/runner.rs [hunk 21]
```diff
@@ -99,6 +112,33 @@ impl<'a> TestCase {
 
         dir.child("src/lib.cairo").touch().unwrap();
 
+        let snforge_std_path = Utf8PathBuf::from_str("../../snforge_std")
+            .unwrap()
+            .canonicalize_utf8()
+            .unwrap()
+            .to_string()
+            .replace('\\', "/");
```
[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/test_utils/src/runner.rs:120]: It is in other 8 places in our codebase too, do we really need that comment?

File: crates/forge/src/scarb.rs [hunk 22]
```diff
@@ -29,10 +31,43 @@ pub fn config_from_scarb_for_package(
...
+        .env("SCARB_PACKAGES_FILTER", filter.to_env())
+        .stderr(Stdio::inherit())
+        .stdout(Stdio::inherit())
+        .output()
+        .context("Failed to build test artifacts with Scarb")?;
+
+    if build_output.status.success() {
+        Ok(())
+    } else {
+        Err(anyhow!("scarb snforge-test-collector did not succeed"))
```
[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/src/scarb.rs:62]: It is a binary call name i.e. `scarb snofrge-test-collector`

File: crates/forge/test_utils/src/runner.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (maciektr) [reply crates/forge/test_utils/src/runner.rs:120]: It doesn't sound like something that should be copied to 8 independent places, but if it is - yes, you need comments, it's beyond confusing.

[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/test_utils/src/runner.rs:120]: This replace happens in other places only in tests to make them work for Windows contributors
